crates/kubelet/src/volumes.rs

```diff
@@ -0,0 +1,186 @@
+        pod: &Pod,
+        client: &kube::Client,
+    ) -> anyhow::Result<HashMap<String, Self>> {
+        let base_path = volume_dir.join(pod_dir_name(pod));
+        tokio::fs::create_dir_all(&base_path).await?;
+        if let Some(vols) = pod.volumes() {
+            let volumes = vols.iter().map(|v| {
+                let mut host_path = base_path.clone();
+                host_path.push(&v.name);
+                async move {
+                    let volume_type = configure(v, pod.namespace(), client, &host_path).await?;
+                    Ok((
+                        v.name.to_owned(),
+                        VolumeRef {
+                            host_path,
+                            volume_type,
+                        },
+                    ))
+                }
+            });
+            futures::future::join_all(volumes)
+                .await
+                .into_iter()
+                .collect()
+        } else {
+            Ok(HashMap::default())
+        }
+    }
+}
+
+impl AsRef<PathBuf> for VolumeRef {
+    fn as_ref(&self) -> &PathBuf {
+        &self.host_path
+    }
+}
+
+impl Deref for VolumeRef {
+    type Target = PathBuf;
+
+    fn deref(&self) -> &Self::Target {
+        &self.host_path
+    }
+}
+
+impl Drop for VolumeRef {
+    fn drop(&mut self) {
+        if matches!(self.volume_type, VolumeType::ConfigMap | VolumeType::Secret) {
+            // TODO: Currently there is no way to do this async (though there is an async destructors proposal)
+            debug!(
+                "deleting {:?} directory {:?}",
+                self.volume_type, self.host_path
+            );
+            std::fs::remove_dir_all(&self.host_path).unwrap_or_else(|e| {
+                error!(
+                    "unable to delete directory {:?} on volume cleanup: {:?}",
+                    self.host_path, e
+                )
+            });
+        }
+    }
+}
+
+/// This is a gnarly function to check all of the supported data members of the Volume struct.
+/// Because it isn't a HashMap, we need to check all fields individually
+async fn configure(
+    vol: &KubeVolume,
+    namespace: &str,
+    client: &kube::Client,
+    path: &PathBuf,
+) -> anyhow::Result<VolumeType> {
+    if let Some(cm) = &vol.config_map {
+        populate_from_config_map(
+            &cm.name
+                .as_ref()
+                .ok_or_else(|| anyhow::anyhow!("no configmap name was given"))?,
+            namespace,
+            client,
+            path,
+        )
+        .await
+    } else if let Some(s) = &vol.secret {
+        populate_from_secret(
+            &s.secret_name
+                .as_ref()
+                .ok_or_else(|| anyhow::anyhow!("no configmap name was given"))?,
+            namespace,
+            client,
+            path,
+        )
+        .await
+    } else {
+        Err(anyhow::anyhow!(
+            "Unsupported volume type. Currently supported types: ConfigMap, Secret, and HostPath"
```

Comment:
the error mentions that it supports the HostPath volume type, yet it isn't checked for and I think a KubeVolume of type hostpath would error out here if I'm not mistaken. Either we should remove HostPath from the supported types or implement `populate_from_hostpath` (unless that isn't necessary, in which case we should noop)

Reply:
So I purposefully left that in here because it is the next piece of work to do and I am probably going to do it in the next day or so. Can remove it if you want